### PR TITLE
mimxrt: Add the pin.irq() method to machine.pin.

### PR DESCRIPTION
**Pull request by robert-hh**
The API complies to the documentation. Both hard and soft irq are supported. Response times for hard irq are <1.5 µs for Teensy40/41 and <3 µs for MIMXRT1020-EVK. Response times for soft irq are <2.6 - 12 µs for Teensy 40/41 and <3.5-38µs for MIMXRT1010-EVK.
Due to the many GPIO ports, the RAM usage is rather high on advantage of response time.

### CODE REVIEW

File: ports/mimxrt/machine_pin.c [hunk 1]
```diff
@@ -234,6 +324,69 @@ STATIC mp_obj_t machine_pin_init(size_t n_args, const mp_obj_t *args, mp_map_t *
...
+    };
+    machine_pin_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // Get the IRQ object.
+    uint32_t gpio_nr = GPIO_get_instance(self->gpio);
+    uint32_t index = GET_PIN_IRQ_INDEX(gpio_nr, self->pin);
+    if (index >= ARRAY_SIZE(MP_STATE_PORT(machine_pin_irq_objects))) {
+        mp_raise_ValueError(MP_ERROR_TEXT("IQR not supported at that PIN"));
```
Comment (dpgeorge) [inline ports/mimxrt/machine_pin.c:343]: "IRQ not supported on given Pin"

File: ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h [hunk 2]
```diff
@@ -8,3 +8,6 @@
 #define MICROPY_HW_LED_ON(pin) (mp_hal_pin_high(pin))
 #define MICROPY_HW_LED_OFF(pin) (mp_hal_pin_low(pin))
 #define BOARD_FLASH_CONFIG_HEADER_H "evkmimxrt1010_flexspi_nor_config.h"
+
+#define MICROPY_BOARD_ROOT_POINTERS \
+    void *machine_pin_irq_objects[2 * 32];
```
Comment (dpgeorge) [inline ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h:13]: would be simpler for a board to define the number of objects, eg `#define MICROPY_HW_NUM_PIN_IRQS (2 * 32)`, and then move this definition into `MICROPY_PORT_ROOT_POINTERS` in `mpconfigport.h`

File: ports/mimxrt/machine_pin.c [hunk 3]
```diff
@@ -68,6 +69,95 @@ const mp_obj_type_t machine_pin_board_pins_obj_type = {
     .locals_dict = (mp_obj_t)&machine_pin_board_pins_locals_dict,
 };
 
+STATIC const mp_irq_methods_t machine_pin_irq_methods;
+
+// STATIC const uint16_t GPIO_irqs[] = GPIO_IRQS;
```
Comment (dpgeorge) [inline ports/mimxrt/machine_pin.c:74]: remove commented out code

File: ports/mimxrt/machine_pin.c [hunk 4]
```diff
@@ -234,6 +324,69 @@ STATIC mp_obj_t machine_pin_init(size_t n_args, const mp_obj_t *args, mp_map_t *
...
+        // Disable all IRQs from the affected source while data is updated.
+        DisableIRQ(irq_num);
+        GPIO_PortDisableInterrupts(self->gpio, 1U << self->pin);
+
+        // Update IRQ data.
+        irq->base.handler = args[ARG_handler].u_obj;
+        irq->base.ishard = args[ARG_hard].u_bool;
+        irq->flags = 0;
+        if (args[ARG_trigger].u_int >= ARRAY_SIZE(IRQ_mapping)) {
+            mp_raise_ValueError(MP_ERROR_TEXT("IQR mode not supported"));
```
Comment (dpgeorge) [inline ports/mimxrt/machine_pin.c:371]: IQR->IRQ

File: ports/mimxrt/pin.h [hunk 5]
```diff
@@ -87,7 +88,14 @@ enum {
...
 
-
+enum {
+    PIN_IRQ_NONE = 0,
+    PIN_IRQ_LOW = 1,
+    PIN_IRQ_HIGH = 2,
+    PIN_IRQ_RISING = 3,
+    PIN_IRQ_FALLING = 4,
+    PIN_IRQ_RISING_OR_FALLING = 0,
+};
```
Comment (dpgeorge) [inline ports/mimxrt/pin.h:98]: looks like this enum can be removed

**Review by dpgeorge — CHANGES_REQUESTED** [review]

File: ports/mimxrt/machine_pin.c (hunk 1 above)
Comment (robert-hh) [reply ports/mimxrt/machine_pin.c:343]: OK. Sounds better.

File: ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h (hunk 2 above)
Comment (robert-hh) [reply ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h:13]: Yes, indeed.

File: ports/mimxrt/machine_pin.c (hunk 3 above)
Comment (robert-hh) [reply ports/mimxrt/machine_pin.c:74]: OK

File: ports/mimxrt/machine_pin.c (hunk 4 above)
Comment (robert-hh) [reply ports/mimxrt/machine_pin.c:371]: That's one one my favorite permutations.

File: ports/mimxrt/pin.h (hunk 5 above)
Comment (robert-hh) [reply ports/mimxrt/pin.h:98]: I have to look into it, but most probably you are right.

**Comment by robert-hh** [comment]
I'll make the changes, test them and supply later today (or your night). Yesterday evening I stumbled over time.time() missing. So I will complete the time module then. All pieces needed  for that are present.

**Comment by dpgeorge** [comment]
Squashed and merged in 7417c6ac91e47798b9c9eba2e7cd8f3dc1bbcc57 with minor edits, including a fix for the default irq trigger (was `kGPIO_IntRisingOrFallingEdge`, changed to 3).

**Pull request closed without merging**